acados/sim/sim_erk_integrator.c

```diff
@@ -616,6 +618,39 @@ int sim_erk(void *config_, sim_in *in, sim_out *out, void *opts_, void *mem_, vo
     ext_fun_arg_t ext_fun_type_out[3];
     void *ext_fun_out[3];
 
+    ext_fun_arg_t expl_vde_type_in[4];
+    void *expl_vde_in[4];
+    ext_fun_arg_t expl_vde_type_out[3];
+    void *expl_vde_out[3];
+
+    int nx_squared_plus_nx = nx * nx + nx;
+    int nx_times_nu = nx * nu;
+
+    if (opts->sens_forw)
+    {  // simulation + forward sensitivities
+        expl_vde_type_in[0] = COLMAJ;
+        expl_vde_in[0] = rhs_forw_in;  // x: nx
+        expl_vde_type_in[1] = COLMAJ;
+        expl_vde_in[1] = rhs_forw_in + nx;  // Sx: nx*nx
+        expl_vde_type_in[2] = COLMAJ;
+        expl_vde_in[2] = rhs_forw_in + nx_squared_plus_nx;  // Su: nx*nu
```

Comment:
Out of curiosity, are these optimization making any difference in performance?
I would expect the bottleneck to be in the function evaluations.

If you get started with such optimizations, you can also consider using a temporary pointer to advance rhs_forw_in, so you avoid to keep performing the same additions.

Reply:
It is hard to measure and I don't remember exactly.
But these computations were inside a double for a triple for loop, so I think the speedup of the ERK change overall is measurable.

Reply:
In the speedup evaluation above, is the casadi function actually having "fake sparse" arguments?
If so, how much does it account for the speedup in the evaluation?

The general feeling is that in ERK everything other than function evaluations is pretty much negligible, and it has been coded in a "textbook" way and the compiler should anyway make a good enough job with that.
If you found this is not the case it would be interesting to investigate if there are other opportunities for code optimization, e.g. also in the floating point algebra.

Reply:
Yes, there are fake sparse arguments.
I ran the same benchmark with a branch excluding the commit [b8c7034386854d6d78f2732e7c6757984a525526](https://github.com/acados/acados/pull/888/commits/b8c7034386854d6d78f2732e7c6757984a525526)
Adding more shooting intervals for more consistent timings.

| Variant | integrators | Remaining Linearization | QP Solver | Remaining QP | Remaining | 
|---------|---------|---------|---------|---------|---------|
| master                      | 26 | 46 | 48 | 34 | 36 |
| dev_c                       | 27 | 36 | 48 | 33 | 11 |
| dev_c_without_fake_sparse   | 32 | 39 | 47 | 34 | 11 |

![acados_FULL_CONDENSING_DAQP_commit_comparison](https://github.com/acados/acados/assets/31900285/458764f8-d61b-48f8-a58b-25e9d23556e6)

Reply:
Ok I see that timings are quite erratic, except there is a clear speedup in "remaining" (that is avoiding the cast of workspace I guess), and to a smaller extent the "fake sparse" casadi, even if master branch is actually faster in this benchmark.

Reply:
Oh right, yes timing is hard :sweat_smile: